auth/scoperesolver.js

```diff
@@ -0,0 +1,338 @@
+      }),
+      // Load all roles on a simplified form: {roleId, scopes}
+      // _computeFixedPoint() will later add the `expandedScopes` property
+      this._Role.scan({}, {
+        handler(role) {
+          let scopes = role.scopes;
+          if (!role.roleId.endsWith('*')) {
+            // Ensure identity, if role isn't a prefix pattern. Obviously,
+            // 'assume:ab' which matches 'assume:a*' doesn't have 'assume:a*'
+            // by the identity relation. But for non-prefix patterns, the
+            // identify relation implies that you have 'assume:<roleId>'.
+            // This speeds up fixed-point computation, and means that if you
+            // have a match without any *, then you can look up the role, and
+            // not have to worry about any prefix patterns that may also match
+            // as they are already saturated.
+            scopes = _.union(scopes, ['assume:' + role.roleId]);
+          }
+          roles.push({roleId: role.roleId, scopes});
+        }
+      })
+    ]);
+
+
+    // Set _roles and _clients at the same time and immediately call
+    // _computeFixedPoint, so anyone using the cache is using a consistent one.
+    this._roles = roles;
+    this._clients = clients;
+    this._computeFixedPoint();
+  }
+
+  /** Compute fixed point over this._roles, and construct _clientCache */
+  _computeFixedPoint() {
+    // Add initial value for expandedScopes for each role R
+    for (let role of this._roles) {
+      role.expandedScopes = _.clone(role.scopes);
+    }
+
+    // Compute fixed-point of roles.expandedScopes for each role R
+    for (let R of this._roles) {
+      let isFixed = false;
+      while (!isFixed) {
+        isFixed = true; // assume we have fixed point
+        for (let role of this._roles) {
+          // if R can assume role, then we union the scope sets, as there is a
+          // finite number of strings here this will reach a fixed-point
+          let test = (scope) => ScopeResolver.grantsRole(scope, role.roleId);
+          if (R.expandedScopes.some(test)) {
+            let count = R.expandedScopes.length;
+            R.expandedScopes = _.union(R.expandedScopes, role.expandedScopes);
+            // Update isFixed, if scopes were added we need a extra round for
+            // proof that we have a fixed point
+            isFixed = isFixed && count == R.expandedScopes.length;
+          }
+        }
+      }
+    }
+
+    // Compress scopes (removing scopes covered by other star scopes)
+    for(let role of this._roles) {
+      role.expandedScopes = ScopeResolver.normalizeScopes(role.expandedScopes);
+    }
+
+    // Construct client cache
+    this._clientCache = {};
+    for (let client of this._clients) {
+      var scopes = this.resolve(['assume:client-id:' + client.clientId]);
+      client.scopes = scopes; // for createSignatureValidator compatibility
+      client.expandedScopes = scopes;
+      this._clientCache[client.clientId] = client;
+    }
```

Comment:
yes,

clientIds are now 1 to 22 chars `[a-z0-9_-]` so we can give human readable clientIds.